File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        # ** Iterative Approach ** #
        # current = root
        # while current:
        #     if p.val < current.val and q.val < current.val:
        #         current = current.left
        #     elif p.val > current.val and q.val > current.val:
        #         current = current.right
        #     else:
        #         return current

        # ** Recursive Approach ** #

        if p.val < root.val and q.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        
        elif p.val > root.val and q.val > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// if p.Val < root.Val && q.Val < root.Val {
    //     return lowestCommonAncestor(root.Left, p, q)
    // }

    // if p.Val > root.Val && q.Val > root.Val {
    //     return lowestCommonAncestor(root.Right, p, q)
    // }

    // return root

    // Iterative Soluion

    cur := root

    for cur != nil {
        if p.Val < cur.Val && q.Val < cur.Val {
            cur = cur.Left
        } else if p.Val > cur.Val && q.Val > cur.Val {
            cur = cur.Right
        } else {
            return cur
        }
    }

    return nil
}